proposals/0174-sbpf-arithmetics-improvements.md

```diff
@@ -0,0 +1,154 @@
+- the `UREM32` instruction (opcode `0x66` and `0x6E`)
+- the `UREM64` instruction (opcode `0x76` and `0x7E`)
+- the `LMUL32` instruction (opcode `0x86` and `0x8E`)
+- the `LMUL64` instruction (opcode `0x96` and `0x9E`)
+- the `SHMUL64` instruction (opcode `0xB6` and `0xBE`)
+- the `SDIV32` instruction (opcode `0xC6` and `0xCE`)
+- the `SDIV64` instruction (opcode `0xD6` and `0xDE`)
+- the `SREM32` instruction (opcode `0xE6` and `0xEE`)
+- the `SREM64` instruction (opcode `0xF6` and `0xFE`)
+
+The verification rule, that an immediate divisor must not be zero or else
+`VerifierError::DivisionByZero` is thrown, is moved to the new division like
+instructions: `UDIV32_IMM`, `UDIV64_IMM`, `UREM32_IMM`, `UREM64_IMM`,
+`SDIV32_IMM`, `SDIV64_IMM`, `SREM32_IMM`, `SREM64_IMM`.
+
+### Changes to Execution
+
+#### PQR Instruction Class
+
+A new instruction class product, quotient and remainder (PQR) is introduced:
+
+- the `UHMUL64` instruction (opcode `0x36` and `0x3E`) produces the 64 MSBs of
+the product of an unsigned 64 x 64 bit multiplication (`dst * imm` and
+`dst * src`).
+- the `UDIV32` instruction (opcode `0x46` and `0x4E`) produces the quotient of
+an unsigned 32 bit division (`dst / imm` and `dst / src`).
+- the `UDIV64` instruction (opcode `0x56` and `0x5E`) produces the quotient of
+an unsigned 64 bit division (`dst / imm` and `dst / src`).
+- the `UREM32` instruction (opcode `0x66` and `0x6E`) produces the remainder of
+an unsigned 64 bit division (`dst % imm` and `dst % src`).
+- the `UREM64` instruction (opcode `0x76` and `0x7E`) produces the remainder of
+an unsigned 64 bit division (`dst % imm` and `dst % src`).
+- the `LMUL32` instruction (opcode `0x86` and `0x8E`) produces the 32 LSBs of
+the product of any 32 x 32 bit multiplication (`dst * imm` and `dst * src`).
+- the `LMUL64` instruction (opcode `0x96` and `0x9E`) produces the 64 LSBs of
+the product of any 64 x 64 bit multiplication (`dst * imm` and `dst * src`).
+- the `SHMUL64` instruction (opcode `0xB6` and `0xBE`) produces the 64 MSBs of
+the product of a signed 64 x 64 bit multiplication (`dst * imm` and
+`dst * src`).
+- the `SDIV32` instruction (opcode `0xC6` and `0xCE`) produces the quotient of
+a signed 32 bit division (`dst / imm` and `dst / src`).
+- the `SDIV64` instruction (opcode `0xD6` and `0xDE`) produces the quotient of
+a signed 64 bit division (`dst / imm` and `dst / src`).
+- the `SREM32` instruction (opcode `0xE6` and `0xEE`) produces the remainder of
+a signed 32 bit division (`dst / imm` and `dst / src`).
+- the `SREM64` instruction (opcode `0xF6` and `0xFE`) produces the remainder of
+a signed 64 bit division (`dst / imm` and `dst / src`).
+
+Runtime exceptions are:
+
+- If the divisor (`imm` or `src`) of any division is zero,
+`EbpfError::DivideByZero` must be thrown.
+- If the dividend (`dst`) of a signed division has the minimal value for its
+bit-width and the divisor (`imm` or `src`) is `-1`, then
+`EbpfError::DivideOverflow` must be thrown.
```

Comment:
Why is it important to error out for this case? We don't error for any other overflows (add, mul, and sub).

Reply:
SDIV64_IMM and SREM64_IMM are exceptions. IMM is 32-bit, so it can't be 64-bit -1. I think it'd be better to remove the check for these.

Reply:
There is sign extension of the inputs (not the outputs). Thus the name "**signed** division".